

Hier sind k-fold cross-validation Scores für die Modelle Logistic Regression, Random Forest und K-Nearest Neighbors wobei die Daten gewichtet wurden. (tödliche bzw tödliche/ schwere Unfälle mit Gewichtung 500)

einmal für tödliche vs. anderer unfälle als zielvariable, ein weiteres mal mit schweren und tödlichen unfällen vs. leichten unfällen als zielvariabe

Mit neuem datensatz, inkl. Lockdown und Covid-Variablen


In [2]:
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold, cross_val_score, cross_val_predict
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

df = pd.read_csv('../data/GeneralDatensatz18-21ohneGeo-mitLockdown_mitCorona.csv', sep=';')

X=df[['UMONAT','USTUNDE','UWOCHENTAG','UART','USTRZUSTAND','BEZ','UTYP1','ULICHTVERH','IstRad','IstPKW','IstFuss','IstKrad','IstGkfz','IstSonstige', 'LOCKDOWN', 'COVID']]
#für tödliche Unfälle
y=(df['UKATEGORIE'] == 1).astype(int)


# KFold-Konfiguration
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Logistische Regression
log_reg = LogisticRegression(max_iter=1000, class_weight={0:1, 1: 500})

accuracy_log = cross_val_score(log_reg, X, y, cv=kf, scoring='accuracy')
# es brauch noch ein X_test und y_test 

pred_log_reg_1 = cross_val_predict(log_reg, X_test, cv=kf)
print("test classification report: ", classification_report(X, pred_log_reg_1))

print("Accuracy Logistic Regression (k-fold):", accuracy_log.mean())


recall_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='recall')
print("Recall Logistic Regression  (k-fold):", recall_reg.mean())

precision_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='precision')
print("Precision Logistic Regression (k-fold):", precision_reg.mean())

f1_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='f1')
print("F1 Logistic Regression (k-fold):", f1_reg.mean())

roc_auc_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='roc_auc')
print("Roc Auc Logistic Regression (k-fold):", roc_auc_reg.mean())


# Random Forest
rf_clas = RandomForestClassifier(n_estimators=100, max_depth=5,random_state=42, class_weight={0:1, 1: 1000})
accuracy_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='accuracy')
print("Accuracy Random Forest (k-fold):", accuracy_rf.mean())

recall_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='recall')
print("Recall Random Forest (k-fold):", recall_rf.mean())

precision_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='precision')
print("Precision Random Forest (k-fold):", precision_rf.mean())

f1_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='f1')
print("F1 Random Forest (k-fold):", f1_rf.mean())

roc_auc_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='roc_auc')
print("Roc Auc Random Forest (k-fold):", roc_auc_rf.mean())


# K-Nearest Neighbors
knn_clas = KNeighborsClassifier(n_neighbors=3, weights='distance')
accuracy_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='accuracy')
print("Accuracy K-Nearest Neighbors (k-fold):", accuracy_knn.mean())

recall_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='recall')
print("Recall K-Nearest Neighbors (k-fold):", recall_knn.mean())

precision_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='precision')
print("Precision K-Nearest Neighbors (k-fold):", precision_knn.mean())

f1_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='f1')
print("F1 K-Nearest Neighbors (k-fold):", f1_knn.mean())

ValueError: Classification metrics can't handle a mix of multiclass-multioutput and binary targets



zB Logistische Regression:

    Recall: 76 % der tödlich Fälle wurden als tödlich erkannt
    Precision: 0,8 % der erkannten tödlich Fälle sind tatsächlich tödlich Fälle ( Precision = tp / (tp + fp) , d.h. wenn es viele False Positive gibt, viele fälschlicherweise als tod klassifiziert werden, dann wird der Nenner klein)



In [2]:
#für tödliche und schwere vs. leichte Unfälle
y = df['UKATEGORIE'].isin([1, 2]).astype(int)
# -> 1 ist schwer oder tödlich, 0 ist leicht

# KFold-Konfiguration
kf = KFold(n_splits=5, shuffle=True, random_state=42)


# Logistische Regression
log_reg = LogisticRegression(max_iter=1000, class_weight={0:1, 1: 500})
accuracy_log = cross_val_score(log_reg, X, y, cv=kf, scoring='accuracy')
print("Accuracy Logistic Regression (k-fold):", accuracy_log.mean())


recall_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='recall')
print("Recall Logistic Regression  (k-fold):", recall_reg.mean())

precision_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='precision')
print("Precision Logistic Regression (k-fold):", precision_reg.mean())

f1_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='f1')
print("F1 Logistic Regression (k-fold):", f1_reg.mean())

roc_auc_reg = cross_val_score(log_reg, X, y, cv=kf, scoring='roc_auc')
print("Roc Auc Logistic Regression (k-fold):", roc_auc_reg.mean())


# Random Forest
rf_clas = RandomForestClassifier(n_estimators=100, max_depth=5,random_state=42, class_weight={0:1, 1: 100})
accuracy_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='accuracy')
print("Accuracy Random Forest (k-fold):", accuracy_rf.mean())

recall_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='recall')
print("Recall Random Forest (k-fold):", recall_rf.mean())

precision_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='precision')
print("Precision Random Forest (k-fold):", precision_rf.mean())

f1_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='f1')
print("F1 Random Forest (k-fold):", f1_rf.mean())

roc_auc_rf = cross_val_score(rf_clas, X, y, cv=kf, scoring='roc_auc')
print("Roc Auc Random Forest (k-fold):", roc_auc_rf.mean())


# K-Nearest Neighbors
knn_clas = KNeighborsClassifier(n_neighbors=3, weights='distance')
accuracy_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='accuracy')
print("Accuracy K-Nearest Neighbors (k-fold):", accuracy_knn.mean())

recall_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='recall')
print("Recall K-Nearest Neighbors (k-fold):", recall_knn.mean())

precision_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='precision')
print("Precision K-Nearest Neighbors (k-fold):", precision_knn.mean())

f1_knn = cross_val_score(knn_clas, X, y, cv=kf, scoring='f1')
print("F1 K-Nearest Neighbors (k-fold):", f1_knn.mean())

Accuracy Logistic Regression (k-fold): 0.1537371408091796
Recall Logistic Regression  (k-fold): 1.0
Precision Logistic Regression (k-fold): 0.1537371408091796
F1 Logistic Regression (k-fold): 0.26650076510709847
Roc Auc Logistic Regression (k-fold): 0.6395680965144174
Accuracy Random Forest (k-fold): 0.1537371408091796
Recall Random Forest (k-fold): 1.0
Precision Random Forest (k-fold): 0.1537371408091796
F1 Random Forest (k-fold): 0.26650076510709847
Roc Auc Random Forest (k-fold): 0.66067902832801
Accuracy K-Nearest Neighbors (k-fold): 0.8091104764707573
Recall K-Nearest Neighbors (k-fold): 0.11347623903069588
Precision K-Nearest Neighbors (k-fold): 0.24217967246100125
F1 K-Nearest Neighbors (k-fold): 0.15443972821930768




zB Logistische Regression:

    Recall: 100 % der schweren/tödlichen Fälle wurden als schweren/tödlichen erkannt
    Precision: 15 % der erkannten schweren/tödlichen Fälle sind tatsächlich schweren/tödlichen Fälle
    ROC AUC Logistic Regression: Der Wert von etwa 0,6393 zeigt, dass das Modell Logistic Regression eine mäßige Fähigkeit hat, zwischen den Klassen zu unterscheiden.

